packages/llamaindex/src/storage/vectorStore/UpstashVectorStore.ts

```diff
@@ -0,0 +1,203 @@
+import {
+  VectorStoreBase,
+  type IEmbedModel,
+  type MetadataFilters,
+  type VectorStoreNoEmbedModel,
+  type VectorStoreQuery,
+  type VectorStoreQueryResult,
+} from "./types.js";
+
+import type { BaseNode, Metadata, TextNode } from "@llamaindex/core/schema";
+import { getEnv } from "@llamaindex/env";
+import { Index } from "@upstash/vector";
+import { metadataDictToNode, nodeToMetadata } from "./utils.js";
+
+type UpstashParams = {
+  namespace?: string;
+  token: string;
+  endpoint: string;
+} & IEmbedModel;
+
+/**
+ * Provides support for writing and querying vector data in Upstash.
+ */
+export class UpstashVectorStore
+  extends VectorStoreBase
+  implements VectorStoreNoEmbedModel
+{
+  storesText: boolean = true;
+
+  private db: Index;
+  namespace: string;
+
+  constructor(params?: UpstashParams) {
+    super(params?.embedModel);
+    this.namespace = params?.namespace ?? "";
+    const token = params?.token ?? getEnv("UPSTASH_VECTOR_REST_TOKEN");
+    const endpoint = params?.endpoint ?? getEnv("UPSTASH_VECTOR_REST_URL");
+
+    if (!token) {
+      throw new Error(
+        "Must specify UPSTASH_VECTOR_REST_TOKEN via env variable.",
+      );
+    }
+    if (!endpoint) {
+      throw new Error("Must specify UPSTASH_VECTOR_REST_URL via env variable.");
+    }
+    this.db = new Index({ token, url: endpoint });
+  }
+
+  private async getDb(): Promise<Index> {
+    if (!this.db) {
+      const { Index } = await import("@upstash/vector");
+      this.db = new Index();
+    }
+
+    return Promise.resolve(this.db);
+  }
+
+  /**
+   * Connects to the database specified in environment vars.
+   * @returns A connection to the database, or the error encountered while connecting/setting up.
+   */
+  client(): Promise<Index> {
+    return this.getDb();
+  }
+
+  /**
+   * Adds vector record(s) to the table.
+   * @param embeddingResults The Nodes to be inserted, optionally including metadata tuples.
+   * @returns Due to limitations in the Upstash client, does not return the upserted ID list, only a Promise resolve/reject.
+   */
+  async add(embeddingResults: BaseNode<Metadata>[]): Promise<string[]> {
+    if (embeddingResults.length == 0) {
+      return Promise.resolve([]);
+    }
+
+    const db = this.db;
+    const nodes = embeddingResults.map(this.nodeToRecord);
+    const result = await db.upsert(nodes, { namespace: this.namespace });
+    if (result != "OK") {
+      return Promise.reject(new Error("Failed to save chunk"));
+    }
+    return Promise.resolve([]);
+  }
+
+  /**
+   * Adds plain text record(s) to the table. Upstash take cares of embedding conversion.
+   * @param embeddingResults The Nodes to be inserted, optionally including metadata tuples.
+   * @returns Due to limitations in the Upstash client, does not return the upserted ID list, only a Promise resolve/reject.
+   */
+  async addPlainText(
```

Comment:
this is already asnyc/await function, we don't need to use `Promise.resolve` / `Promise.reject`. Just return to throw.